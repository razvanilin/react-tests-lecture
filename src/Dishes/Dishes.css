#searchBar{
  margin-bottom: 20px;
  margin-left: 0.2%;
}
