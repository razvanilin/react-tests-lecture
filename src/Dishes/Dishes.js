import React, { Component } from "react";
// Alternative to passing the moderl as the component property,
// we can import the model instance directly
import modelInstance from "../data/DinnerModel";
import { Link } from "react-router-dom";
import "./Dishes.css";

class Dishes extends Component {
  constructor(props) {
    super(props);
    // We create the state to store the various statuses
    // e.g. API data loading or error
    this.state = {
      status: "LOADING",
      filter: "",
      type: "All"
    };
    this.changeFilter = this.changeFilter.bind(this);
    this.changeType = this.changeType.bind(this);
    //this.searchUpdate = this.searchUpdate.bind(this);
  }

  // this methods is called by React lifecycle when the
  // component is actually shown to the user (mounted to DOM)
  // that's a good place to call the API and get the data
  async componentDidMount() {
    await this.searchUpdate();
  }

  changeFilter(filterInput){
    this.setState({
      filter: filterInput.target.value
    })
  }

  changeType(typeInput){
    this.setState({
      type: typeInput.target.value
    })
  }

  searchUpdate = () => {
    modelInstance
      .getAllDishes(this.state.type, this.state.filter)
      .then(dishes => {
        this.setState({
          status: "LOADED",
          dishes: dishes.results
        });
      })
      .catch(() => {
        this.setState({
          status: "ERROR"
        });
      });
  }

  render() {
    let dishesList = null;
    const { filter } = this.state;

    // depending on the state we either generate
    // useful message to the user or show the list
    // of returned dishes
    switch (this.state.status) {
      case "LOADING":
        dishesList = <em>Loading...</em>;
        break;
      case "LOADED":
        dishesList = this.state.dishes.map(dish => (
            <div className="col-lg-3 col-xs-12 col-sm-12 text-center" key={dish.id}>
              <Link to={"/dishDetail/"+dish.id}>
                <img src={`http://spoonacular.com/recipeImages/${dish.image}`} alt='Food' height="200" width="200"/>
              </Link>
              <p>{dish.title}</p>
            </div>
        ));
        break;
      default:
        dishesList = <b>Failed to load data, please try again</b>;
        break;
    }

    return (
      <div className="Dishes">
        <h3>FIND A DISH</h3>
        <div className="row" id="searchBar">
            <input id="searchText" type="text" placeholder="Enter keyword..." value={filter} onChange={this.changeFilter}/>
            <select id="searchType" onChange={this.changeType}>
              <option value="All">All</option>
              <option value="main course">Main Course</option>
              <option value="side dish">Side Dish</option>
              <option value="dessert">Dessert</option>
              <option value="appetizer">Appetizer</option>
              <option value="salad">Salad</option>
              <option value="bread">Bread</option>
              <option value="soup">Soup</option>
              <option value="beverage">Beverage</option>
              <option value="sauce">Sauce</option>
              <option value="drink">Drink</option>
              <option value="breakfast">Breakfast</option>
            </select>
          <div id="searchBarButton">
            <button type="button" className="btn btn-warning" id="searchButton" onClick={this.searchUpdate}>Search</button>
          </div>
        </div>
        <div className="row">{dishesList}</div>
      </div>
    );
  }
}

export default Dishes;
