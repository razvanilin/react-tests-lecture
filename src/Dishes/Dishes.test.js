import React from "react";
import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import { mount, shallow } from "enzyme";

import Dishes from "./Dishes";
import dinnerModel from "../data/DinnerModel";

it("renders the Dishes correctly", () => {
  const tree = shallow(
    <MemoryRouter>
      <Dishes />
    </MemoryRouter>
  );

  expect(tree.find(Dishes)).toMatchSnapshot();
  tree.unmount();
});

it("checks the onChange filter function and state changes", () => {
  const DishesComponent = shallow(
    <MemoryRouter>
      <Dishes />
    </MemoryRouter>
  );

  const filterInput = DishesComponent.find(Dishes).dive().find("input");

  filterInput.simulate("change", { target: { value: "vanilla" } });
  setTimeout(() => {
    expect(DishesComponent.state().filter).toBe("vanilla");
    DishesComponent.unmount();
  });
});
