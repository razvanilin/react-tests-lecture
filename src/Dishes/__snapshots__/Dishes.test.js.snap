// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Dishes correctly 1`] = `<Dishes />`;
